File: swift/Sources/SignalClient/zkgroup/ReceiptCredentialPresentation.swift
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import SignalFfi

public class ReceiptCredentialPresentation: ByteArray {

  public required init(contents: [UInt8]) throws {
    try super.init(contents, checkValid: signal_receipt_credential_presentation_check_valid_contents)
  }

  public func getReceiptExpirationTime() throws -> UInt64 {
    return try withUnsafePointerToSerialized { contents in
      try invokeFnReturningInteger {
        signal_receipt_credential_presentation_get_receipt_expiration_time($0, contents)
      }
    }
  }

  public func getReceiptLevel() throws -> UInt64 {
    return try withUnsafePointerToSerialized { contents in
      try invokeFnReturningInteger {
        signal_receipt_credential_presentation_get_receipt_level($0, contents)
      }
    }
  }

  public func getReceiptSerial() throws -> ReceiptSerial {
    return try withUnsafePointerToSerialized { contents in
      try invokeFnReturningSerialized {
        signal_receipt_credential_presentation_get_receipt_serial($0, contents)
      }
    }
  }

}

File: java/java/src/main/java/org/signal/zkgroup/receipts/ReceiptCredentialPresentation.java
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.zkgroup.receipts;

import java.nio.ByteBuffer;
import org.signal.zkgroup.InvalidInputException;
import org.signal.zkgroup.internal.ByteArray;
import org.signal.client.internal.Native;

public final class ReceiptCredentialPresentation extends ByteArray {
  public ReceiptCredentialPresentation(byte[] contents) throws InvalidInputException {
    super(contents);
    Native.ReceiptCredentialPresentation_CheckValidContents(contents);
  }

  public long getReceiptExpirationTime() {
    return Native.ReceiptCredentialPresentation_GetReceiptExpirationTime(contents);
  }

  public long getReceiptLevel() {
    return Native.ReceiptCredentialPresentation_GetReceiptLevel(contents);
  }

  public ReceiptSerial getReceiptSerial() {
    byte[] newContents = Native.ReceiptCredentialPresentation_GetReceiptSerial(contents);

    try {
      return new ReceiptSerial(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

}
